File: cses/09_geometry/2192_point-in-polygon/solution.py
import sys


def solve() -> None:
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    # Toạ độ các đỉnh đa giác theo thứ tự.
    xs = [0] * n
    ys = [0] * n
    for i in range(n):
        xs[i] = int(data[idx]); idx += 1
        ys[i] = int(data[idx]); idx += 1

    out = []
    for _ in range(m):
        px = int(data[idx]); idx += 1
        py = int(data[idx]); idx += 1

        inside = False
        on_boundary = False

        # Duyệt từng cạnh (x1,y1)-(x2,y2) của đa giác.
        x1 = xs[n - 1]
        y1 = ys[n - 1]
        for i in range(n):
            x2 = xs[i]
            y2 = ys[i]

            # --- BOUNDARY: điểm có nằm trên cạnh (x1,y1)-(x2,y2) không? ---
            # cross product == 0 nghĩa là ba điểm thẳng hàng; sau đó kiểm tra
            # điểm nằm trong bounding box của cạnh.
            cross = (x2 - x1) * (py - y1) - (y2 - y1) * (px - x1)
            if cross == 0:
                if min(x1, x2) <= px <= max(x1, x2) and \
                   min(y1, y2) <= py <= max(y1, y2):
                    on_boundary = True
                    break

            # --- RAY CASTING: bắn tia về hướng +x và đếm số lần cắt. ---
            # Half-open trick (y1 > py) != (y2 > py) để mỗi đỉnh chỉ đếm một lần.
            if (y1 > py) != (y2 > py):
                dy = y2 - y1
                # Cạnh cắt đường ngang y = py tại hoành độ:
                #   xint = x1 + (x2-x1)*(py-y1)/(y2-y1)
                # Cần biết px < xint (giao điểm nằm bên phải P). Nhân chéo để
                # tránh số thực, đồng thời xét dấu dy.
                lhs = (px - x1) * dy               # (px-x1)*(y2-y1)
                rhs = (x2 - x1) * (py - y1)         # (x2-x1)*(py-y1)
                if dy > 0:
                    if lhs < rhs:
                        inside = not inside
                else:  # dy < 0
                    if lhs > rhs:
                        inside = not inside

            x1 = x2
            y1 = y2

        # Số lần cắt lẻ -> INSIDE, chẵn -> OUTSIDE; ưu tiên BOUNDARY.
        if on_boundary:
            out.append("BOUNDARY")
        elif inside:
            out.append("INSIDE")
        else:
            out.append("OUTSIDE")

    sys.stdout.write("\n".join(out) + "\n")


solve()

File: cses/09_geometry/2192_point-in-polygon/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

struct Point {
    long long x;
    long long y;
};

// Tích có hướng (cross product) của vector a->b với vector a->p.
// Bằng 0 nghĩa là ba điểm thẳng hàng.
long long cross(const Point& a, const Point& b, const Point& p) {
    return (b.x - a.x) * (p.y - a.y) -
           (b.y - a.y) * (p.x - a.x);
}

// P nằm trên đoạn a-b khi thẳng hàng (cross == 0) và nằm trong bounding box.
bool lies_on_segment(const Point& a, const Point& b, const Point& p) {
    return cross(a, b, p) == 0 &&
           std::min(a.x, b.x) <= p.x && p.x <= std::max(a.x, b.x) &&
           std::min(a.y, b.y) <= p.y && p.y <= std::max(a.y, b.y);
}

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;
    std::vector<Point> polygon(static_cast<std::size_t>(n));
    for (Point& point : polygon) {
        std::cin >> point.x >> point.y;
    }

    for (int query = 0; query < m; ++query) {
        Point p{};
        std::cin >> p.x >> p.y;

        bool inside = false;
        bool boundary = false;
        Point a = polygon.back();

        // Duyệt từng cạnh a->b của đa giác.
        for (const Point& b : polygon) {
            // Ưu tiên kiểm tra BOUNDARY: điểm nằm trên cạnh thì dừng ngay.
            if (lies_on_segment(a, b, p)) {
                boundary = true;
                break;
            }

            // Ray casting theo hướng +x. Half-open trick: khoảng nửa mở theo y
            // giúp mỗi đỉnh chung chỉ được đếm đúng một lần.
            if ((a.y > p.y) != (b.y > p.y)) {
                // Nhân chéo thay cho phép chia để giữ số học nguyên chính xác;
                // xét dấu dy để so px với hoành độ giao điểm.
                const long long dy = b.y - a.y;
                const long long left = (p.x - a.x) * dy;
                const long long right = (b.x - a.x) * (p.y - a.y);
                if ((dy > 0 && left < right) ||
                    (dy < 0 && left > right)) {
                    inside = !inside;
                }
            }
            a = b;
        }

        // Số lần cắt lẻ -> INSIDE, chẵn -> OUTSIDE.
        if (boundary) {
            std::cout << "BOUNDARY\n";
        } else if (inside) {
            std::cout << "INSIDE\n";
        } else {
            std::cout << "OUTSIDE\n";
        }
    }
    return 0;
}
